File: frontend/src/api/marketing.ts
import { api } from './client'

export type SemanticMode = 'seo' | 'direct'

export interface SemanticProject {
  id: string
  project_id: string
  name: string
  mode: SemanticMode
  region: string | null
  region_id: number | null
  is_seasonal: boolean
  needs_brand_check: boolean
  pipeline_step: number
  created_at: string
  updated_at: string
}

export interface SemanticKeyword {
  id: string
  phrase: string
  frequency_base: number | null
  frequency_phrase: number | null
  frequency_exact: number | null
  frequency_order: number | null
  kw_type: 'ВЧ' | 'СЧ' | 'НЧ' | null
  intent: string | null
  source: string
  is_mask: boolean
  mask_selected: boolean
  is_branded: boolean
  is_competitor: boolean
  is_seasonal: boolean
  geo_dependent: boolean
  is_excluded: boolean
  cluster_name: string | null
  created_at: string
}

export interface SemanticCluster {
  id: string
  name: string
  intent: string | null
  priority: string | null
  campaign_type: string | null
  suggested_title: string | null
  suggested_description: string | null
  keyword_count: number
  created_at: string
  updated_at: string
}

export interface MinusWord {
  id: string
  word: string
  note: string | null
  added_at: string
}

export interface KeywordsListResponse {
  items: SemanticKeyword[]
  total: number
  page: number
  per_page: number
}

export interface KeywordsParams {
  page?: number
  per_page?: number
  kw_type?: string
  intent?: string
  source?: string
  only_masks?: boolean
  search?: string
}

export const marketingApi = {
  // ── Semantic Projects ───────────────────────────────────────────────────────

  list: (projectId: string): Promise<SemanticProject[]> =>
    api.get(`/projects/${projectId}/marketing/semantic`).then((r) => r.data),

  get: (projectId: string, semId: string): Promise<SemanticProject> =>
    api.get(`/projects/${projectId}/marketing/semantic/${semId}`).then((r) => r.data),

  create: (
    projectId: string,
    data: { name: string; mode: SemanticMode; region?: string | null; region_id?: number | null; is_seasonal?: boolean }
  ): Promise<SemanticProject> =>
    api.post(`/projects/${projectId}/marketing/semantic`, data).then((r) => r.data),

  delete: (projectId: string, semId: string): Promise<void> =>
    api.delete(`/projects/${projectId}/marketing/semantic/${semId}`).then(() => undefined),

  // ── Autopilot ──────────────────────────────────────────────────────────────

  autopilot: (
    projectId: string,
    semId: string,
    data: { min_freq_exact: number }
  ): Promise<{ task_id: string; status: string }> =>
    api.post(`/projects/${projectId}/marketing/semantic/${semId}/autopilot`, data).then((r) => r.data),

  // ── Masks ───────────────────────────────────────────────────────────────────

  expand: (
    projectId: string,
    semId: string,
    data: { min_freq_exact: number; use_brief: boolean }
  ): Promise<{ task_id: string; status: string }> =>
    api.post(`/projects/${projectId}/marketing/semantic/${semId}/expand`, data).then((r) => r.data),

  collectMasks: (projectId: string, semId: string, masks: string[]): Promise<SemanticKeyword[]> =>
    api
      .post(`/projects/${projectId}/marketing/semantic/${semId}/collect-masks`, { masks })
      .then((r) => r.data),

  updateMaskSelection: (
    projectId: string,
    semId: string,
    kwId: string,
    mask_selected: boolean
  ): Promise<SemanticKeyword> =>
    api
      .patch(
        `/projects/${projectId}/marketing/semantic/${semId}/keywords/${kwId}/mask-selection`,
        { mask_selected }
      )
      .then((r) => r.data),

  // ── Keywords ─────────────────────────────────────────────────────────────────

  getKeywords: (projectId: string, semId: string, params?: KeywordsParams): Promise<KeywordsListResponse> =>
    api
      .get(`/projects/${projectId}/marketing/semantic/${semId}/keywords`, { params })
      .then((r) => r.data),

  updateKeyword: (
    projectId: string,
    semId: string,
    kwId: string,
    data: {
      is_excluded?: boolean
      is_branded?: boolean
      is_competitor?: boolean
      is_seasonal?: boolean
      geo_dependent?: boolean
      intent?: string
    }
  ): Promise<SemanticKeyword> =>
    api
      .patch(`/projects/${projectId}/marketing/semantic/${semId}/keywords/${kwId}`, data)
      .then((r) => r.data),

  // ── Cleaning ──────────────────────────────────────────────────────────────────

  autoClean: (
    projectId: string,
    semId: string
  ): Promise<{
    excluded_zero_freq: number
    excluded_long_tail: number
    excluded_minus_words: number
    total_excluded: number
    total_kept: number
    snapshot_id: string
  }> => api.post(`/projects/${projectId}/marketing/semantic/${semId}/auto-clean`).then((r) => r.data),

  completeCleaning: (projectId: string, semId: string): Promise<SemanticProject> =>
    api.post(`/projects/${projectId}/marketing/semantic/${semId}/cleaning/complete`).then((r) => r.data),

  // ── Minus Words ───────────────────────────────────────────────────────────────

  getMinusWords: (projectId: string, semId: string): Promise<MinusWord[]> =>
    api.get(`/projects/${projectId}/marketing/semantic/${semId}/minus-words`).then((r) => r.data),

  addMinusWords: (projectId: string, semId: string, words: string[]): Promise<MinusWord[]> =>
    api
      .post(`/projects/${projectId}/marketing/semantic/${semId}/minus-words`, words.map((w) => ({ word: w })))
      .then((r) => r.data),

  deleteMinusWord: (projectId: string, semId: string, wordId: string): Promise<void> =>
    api
      .delete(`/projects/${projectId}/marketing/semantic/${semId}/minus-words/${wordId}`)
      .then(() => undefined),

  // ── Clustering ─────────────────────────────────────────────────────────────

  startCluster: (projectId: string, semId: string): Promise<{ task_id: string; status: string }> =>
    api.post(`/projects/${projectId}/marketing/semantic/${semId}/cluster`).then((r) => r.data),

  getClusters: (projectId: string, semId: string): Promise<SemanticCluster[]> =>
    api.get(`/projects/${projectId}/marketing/semantic/${semId}/clusters`).then((r) => r.data),

  updateCluster: (
    projectId: string,
    semId: string,
    clusterId: string,
    data: Partial<Pick<SemanticCluster, 'name' | 'intent' | 'priority' | 'campaign_type' | 'suggested_title' | 'suggested_description'>>
  ): Promise<SemanticCluster> =>
    api
      .patch(`/projects/${projectId}/marketing/semantic/${semId}/clusters/${clusterId}`, data)
      .then((r) => r.data),

  deleteCluster: (projectId: string, semId: string, clusterId: string): Promise<void> =>
    api
      .delete(`/projects/${projectId}/marketing/semantic/${semId}/clusters/${clusterId}`)
      .then(() => undefined),

  // ── Export ───────────────────────────────────────────────────────────────────

  exportUrl: (projectId: string, semId: string, fmt: 'xlsx' | 'csv' | 'txt'): string =>
    `/api/projects/${projectId}/marketing/semantic/${semId}/export?fmt=${fmt}`,

  exportBlob: (projectId: string, semId: string, fmt: 'xlsx' | 'csv' | 'txt'): Promise<{ blob: Blob; filename: string }> =>
    api
      .get(`/projects/${projectId}/marketing/semantic/${semId}/export`, {
        params: { fmt },
        responseType: 'blob',
      })
      .then((r) => {
        const disposition: string = r.headers['content-disposition'] ?? ''
        const match = disposition.match(/filename="([^"]+)"/)
        const filename = match ? match[1] : `semantic.${fmt}`
        return { blob: r.data as Blob, filename }
      }),
}

File: backend/app/models/marketing.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import JSON, Boolean, DateTime, Enum, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin


class SemanticMode(str, enum.Enum):
    SEO = "seo"
    DIRECT = "direct"


class SemanticProject(Base, TimestampMixin):
    __tablename__ = "semantic_projects"
    __table_args__ = (
        UniqueConstraint("project_id", "mode", name="uq_semantic_project_mode"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    mode: Mapped[SemanticMode] = mapped_column(
        Enum(SemanticMode, values_callable=lambda x: [e.value for e in x]), nullable=False
    )
    region: Mapped[str | None] = mapped_column(String(100), nullable=True)
    region_id: Mapped[int | None] = mapped_column(Integer, nullable=True)  # Yandex region ID
    is_seasonal: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    brand_check_date: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    needs_brand_check: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # 0=created, 1=masks done, 2=expanded, 3=cleaned, 4=clustered
    pipeline_step: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    config: Mapped[dict | None] = mapped_column(JSON, nullable=True)  # niche template overrides, competitor crawl cache


class SemanticKeyword(Base, TimestampMixin):
    __tablename__ = "semantic_keywords"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    semantic_project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("semantic_projects.id"), nullable=False, index=True
    )

    phrase: Mapped[str] = mapped_column(String(512), nullable=False)

    # 4 Wordstat frequency types
    frequency_base: Mapped[int | None] = mapped_column(Integer, nullable=True)    # WS
    frequency_phrase: Mapped[int | None] = mapped_column(Integer, nullable=True)  # «WS»
    frequency_exact: Mapped[int | None] = mapped_column(Integer, nullable=True)   # «!WS»
    frequency_order: Mapped[int | None] = mapped_column(Integer, nullable=True)   # [WS]

    kw_type: Mapped[str | None] = mapped_column(String(10), nullable=True)    # ВЧ | СЧ | НЧ
    intent: Mapped[str | None] = mapped_column(String(50), nullable=True)
    source: Mapped[str] = mapped_column(String(50), nullable=False, default="wordstat")

    geo_dependent: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_branded: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_seasonal: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_competitor: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    is_excluded: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    excluded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    cluster_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    monthly_dynamics: Mapped[dict | None] = mapped_column(JSON, nullable=True)

    # Mask-specific
    is_mask: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    mask_selected: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)


class KeywordCache(Base):
    __tablename__ = "keyword_cache"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    phrase: Mapped[str] = mapped_column(String(512), nullable=False, index=True)
    region_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frequency_base: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frequency_phrase: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frequency_exact: Mapped[int | None] = mapped_column(Integer, nullable=True)
    frequency_order: Mapped[int | None] = mapped_column(Integer, nullable=True)
    monthly_dynamics: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    cached_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class CleaningSnapshot(Base):
    __tablename__ = "cleaning_snapshots"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    semantic_project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("semantic_projects.id"), nullable=False, index=True
    )
    snapshot: Mapped[list] = mapped_column(JSON, nullable=False)
    description: Mapped[str] = mapped_column(String(255), nullable=False, default="авто-очистка")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class SemanticCluster(Base, TimestampMixin):
    __tablename__ = "semantic_clusters"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    semantic_project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("semantic_projects.id"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    intent: Mapped[str | None] = mapped_column(String(50), nullable=True)
    priority: Mapped[str | None] = mapped_column(String(20), nullable=True)
    campaign_type: Mapped[str | None] = mapped_column(String(50), nullable=True)  # search | rsa (Direct mode)
    related_cluster_ids: Mapped[list | None] = mapped_column(JSON, nullable=True)
    suggested_title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    suggested_description: Mapped[str | None] = mapped_column(Text, nullable=True)


class MarketingMinusWord(Base):
    __tablename__ = "marketing_minus_words"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    semantic_project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("semantic_projects.id"), nullable=False, index=True
    )
    word: Mapped[str] = mapped_column(String(255), nullable=False)
    note: Mapped[str | None] = mapped_column(String(255), nullable=True)
    added_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
